kernel/src/expressions/single_row_transform.rs

```diff
@@ -0,0 +1,540 @@
+        if let Some(err) = &self.error.replace(e) {
+            debug!("Overwriting error that was already set: {err}");
+        }
+    }
+
+    fn check_error<T, E: Into<Error>>(&mut self, result: Result<T, E>) -> Option<T> {
+        match result {
+            Ok(val) => Some(val),
+            Err(err) => {
+                self.set_error(err.into());
+                None
+            }
+        }
+    }
+}
+
+impl<'a, T: Iterator<Item = &'a Scalar>> SchemaTransform<'a> for SingleRowTransform<'a, T> {
+    fn transform_primitive(
+        &mut self,
+        prim_type: &'a PrimitiveType,
+    ) -> Option<Cow<'a, PrimitiveType>> {
+        // first always check error to terminate early if possible
+        if self.error.is_some() {
+            return None;
+        }
+
+        let next = self.scalars.next();
+        let scalar = self.check_error(next.ok_or(Error::InsufficientScalars))?;
+
+        let DataType::Primitive(scalar_type) = scalar.data_type() else {
+            self.set_error(Error::Schema(
+                "Non-primitive scalar type {datatype} provided".to_string(),
+            ));
+            return None;
+        };
+        if scalar_type != *prim_type {
+            self.set_error(Error::Schema(format!(
+                "Mismatched scalar type creating a single-row array: expected {}, got {}",
+                prim_type, scalar_type
+            )));
+            return None;
+        }
+
+        self.stack.push(Expression::Literal(scalar.clone()));
+        Some(Cow::Borrowed(prim_type))
+    }
+
+    fn transform_struct(&mut self, struct_type: &'a StructType) -> Option<Cow<'a, StructType>> {
+        // first always check error to terminate early if possible
+        if self.error.is_some() {
+            return None;
+        }
+
+        // Only consume newly-added entries (if any). There could be fewer than expected if
+        // the recursion encountered an error.
+        let mark = self.stack.len();
+        self.recurse_into_struct(struct_type)?;
+        let field_exprs = self.stack.split_off(mark);
+
+        if field_exprs.len() != struct_type.fields_len() {
+            self.set_error(Error::InsufficientScalars);
+            return None;
+        }
+
+        let mut found_non_nullable_null = false;
+        let mut all_null = true;
+        let fields = struct_type.fields();
+        for (field, expr) in fields.zip(&field_exprs) {
+            if !matches!(expr, Expression::Literal(Scalar::Null(_))) {
+                all_null = false;
+            } else if !field.is_nullable() {
+                found_non_nullable_null = true;
+            }
+        }
+
+        // If all children are NULL and at least one is ostensibly non-nullable, we interpret
+        // the struct itself as being NULL
+        let struct_expr = if all_null && found_non_nullable_null {
+            Expression::null_literal(struct_type.clone().into())
+        } else {
+            Expression::struct_from(field_exprs)
```

Comment:
don't we need to ensure that no field is actually null that wasn't marked as nullable?

Reply:
yep added!